### PR TITLE
Migrate WebSocket transport networking

### PR DESCRIPTION
**Pull request by lsm**
## Summary
- Route WebSocket connection setup and stream read/write through Makai's `compat.net` networking seam for the Zig 0.16.0 migration.
- Keep nonce and masking key generation on the existing `compat.random.fillSecureBytes` wrapper.
- Document that `Sec-WebSocket-Accept` value verification is deferred, `wss://` TLS remains unsupported, and `std.Io.Evented` networking is not supported by the current seam.

## Context
Zig 0.15.2 → 0.16.0 Migration — WebSocket transport networking

## Test plan
- [x] `cd zig && zig build test-unit-transport`
- [x] `cd zig && zig build test`

### CODE REVIEW

File: zig/src/transports/websocket.zig [hunk 1]
```diff
@@ -622,12 +613,12 @@ fn performHandshake(
...
-    // For simplicity, just verify the header exists
-    if (std.mem.find(u8, response, "Sec-WebSocket-Accept:") == null and
-        std.mem.find(u8, response, "Sec-WebSocket-Protocol:") == null)
-    {
+    // Verify Sec-WebSocket-Accept header.
+    // The expected accept key is base64(sha1(key ++
+    // "258EAFA5-E914-47DA-95CA-C5AB0DC85B11")). Full value verification is
+    // intentionally deferred; this migration preserves existing behavior by
+    // only requiring the accept header to be present.
+    if (std.mem.find(u8, response, "Sec-WebSocket-Accept:") == null) {
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/transports/websocket.zig]: Fixed in ede3ae7 by parsing response header lines and comparing header names with std.ascii.eqlIgnoreCase. I also added a regression test covering lowercase sec-websocket-accept and other case-insensitive header names.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by claude-opus-4-20250514 (Anthropic)

> **Model:** claude-opus-4-20250514 | **Client:** NeoKai | **Provider:** Anthropic

## Summary

This PR successfully migrates the WebSocket transport from Zig 0.15.2 to 0.16.0 networking APIs through Makai's compatibility seams. The migration preserves transport interface compatibility, correctly routes networking through `compat.net`, and properly maintains security-sensitive random byte generation through `compat.random.fillSecureBytes`.

## ✅ Approval Criteria Met

All acceptance criteria from the task description are satisfied:

- ✅ **Compilation**: `websocket.zig` compiles successfully on Zig 0.16.0
- ✅ **Tests pass**: All transport unit tests pass (`zig build test-unit-transport`), full test suite passes
- ✅ **Documentation added**: File header documents `Sec-WebSocket-Accept` value verification is deferred, `wss://` TLS is unsupported, and `std.Io.Evented` networking is not supported
- ✅ **Bug fix addressed**: Case-insensitive header matching implemented in commit `ede3ae7` with regression test

## Technical Analysis

### ✅ Networking Migration (Correct)

The migration correctly routes all networking operations through `compat.net`:

- **Connection setup** (line 107): Uses `compat.net.tcpConnectHost()` for address resolution and TCP connection
- **Stream I/O** (lines 39-57, 145, 198, 230): `compat.net.Stream` wrapper provides `read()`, `write()`, `writeAll()`, `close()` methods
- **Socket handle conversion** (lines 21-23): `streamFromSocketHandle()` properly wraps raw handles for test infrastructure

The implementation maintains the same public `WebSocketClient` API surface while hiding `std.Io` backend details behind the compatibility seam.

### ✅ Random/Masking Key Generation (Correct)

Security-sensitive nonce and masking key generation correctly uses `compat.random.fillSecureBytes()`:

- **Handshake nonce** (line 580): `compat.random.fillSecureBytes(&nonce)` for 16-byte WebSocket key
- **Frame masking** (line 484): `compat.random.fillSecureBytes(&mask)` for XOR mask

This preserves the security boundary while abstracting away `std.Io.randomSecure`.

### ✅ Transport Interface Implementation (Correct)

`WebSocketClient` correctly implements both required transport interfaces:

- **`AsyncSender`** (lines 243-250): Provides `write_fn`, `flush_fn`, `close_fn`
- **`AsyncReceiver`** (lines 253-260): Provides `receive_stream_fn`, `read_fn`, `close_fn`

Memory ownership is properly handled:
- Line 187: `receive()` returns `allocator.dupe(u8, frame.payload)` (owned allocation)
- Line 357: Byte chunk correctly marked with `owned = true`
- Lines 345-348: ProducerContext cleanup properly handles both success and error paths

### ✅ Error Handling (Correct)

Error types remain consistent:
- `error.TlsNotSupported` (line 102): Properly rejects `wss://` before networking
- `error.ConnectionFailed` (line 109): Wrapped from `compat.net.tcpConnectHost`
- `error.HandshakeFailed` (line 642): Validates response status/upgrade/accept headers
- State management resets to `.disconnected` on all error paths (lines 115, 120)

### ✅ Case-Insensitive Header Matching (Fixed)

Commit `ede3ae7` correctly addresses the P2 feedback about HTTP header case-insensitivity:

- **`hasHeaderName()` function** (lines 388-406): Parses response lines and compares names with `std.ascii.eqlIgnoreCase()`
- **Regression test** (lines 766-777): Verifies lowercase `sec-websocket-accept`, `upgrade`, `connection` are matched

This fixes the interoperability issue with servers that normalize header names to lowercase.

### ✅ Test Coverage (Adequate for Scope)

The PR includes 23 unit tests covering:
- Frame encoding/decoding (7 tests)
- Client lifecycle management (8 tests)
- Protocol validation (3 tests)
- Interface conformance (1 test)
- Edge cases (4 tests)

**Note**: As noted in the migration plan, WebSocket transport has no E2E or protocol-level integration tests. This is acceptable for the current scope (development/testing use), as the transport is not yet used in production protocol/agent layers. The unit tests sufficiently validate the networking migration.

### ✅ Documentation (Complete)

File header (lines 1-14) clearly documents:
- `wss://` TLS limitation: parsed but rejected with `error.TlsNotSupported`
- Networking backend: `std.Io.Threaded` context used, `std.Io.Evented` unsupported
- Handshake validation: checks presence of `Sec-WebSocket-Accept`, not value verification

## 🎯 Conclusion

**APPROVE** — This is a clean, well-executed migration that correctly implements the Zig 0.16.0 networking APIs through Makai's compatibility seams. The code maintains interface compatibility, properly handles error paths, and includes adequate unit test coverage. The documentation clearly outlines the known limitations (TLS, accept value verification, Evented networking), which are explicitly deferred per the migration plan.

The PR is ready to merge.
